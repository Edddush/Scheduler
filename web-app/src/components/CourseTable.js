import React from "react";
import Table from "@mui/material/Table";
import TableBody from "@mui/material/TableBody";
import TableCell from "@mui/material/TableCell";
import TableContainer from "@mui/material/TableContainer";
import TableHead from "@mui/material/TableHead";
import TableRow from "@mui/material/TableRow";
import Paper from "@mui/material/Paper";
import Button from "react-bootstrap/Button";
import uuid from "react-uuid";

//-----------------------------------------------------------------------------------------------------------------------------
// Utility table to display x rows with y columns of data where the last column is a callback relating to the data

// props structure:

// data is an array of objects, where each object contains the properties defined in the columns object
// columns is an array of objects, each object has a property called header which will be the display text,
// each object also has a property called accessProperty which is the name of the property on the objects in data that
// corresponds to the header for that column
// callbackCol represents the last column and is an object with three properties -> header, callback, and text
//-----------------------------------------------------------------------------------------------------------------------------

function CourseTable({ data: rows, columns, callbackCol }) {
  return (
    <TableContainer
      component={Paper}
      style={{
        height: "400px",
        overflowY: "scroll",
      }}
    >
      <Table
        stickyHeader
        sx={{ minWidth: 650 }}
        style={{
          overflow: "scroll",
        }}
        aria-label="simple table"
      >
        <TableHead>
          <TableRow>
            {columns.map(({header}) => (
              <TableCell key={uuid()} align="center">
                {header}
              </TableCell>
            ))}
            <TableCell align="center">{callbackCol.header}</TableCell>
          </TableRow>
        </TableHead>
        <TableBody>
          {rows.map((row,i) => (
            <TableRow
              key={`${row.name}${i}`}
              sx={{ "&:last-child td, &:last-child th": { border: 0 } }}
            >
              {
                // map through cols and get data for that row
                columns.map((col) => (
                  <TableCell key={uuid()} align="center">
                    {row[col.accessProperty]}
                  </TableCell>
                ))
              }
              <TableCell align="center">
                <Button
                  id="temp-button"
                  variant="dark"
                  size="lg"
                  onClick={() => callbackCol.callback(row)}
                >
                  {callbackCol.text}
                </Button>
              </TableCell>
            </TableRow>
          ))}
        </TableBody>
      </Table>
    </TableContainer>
  );
}

export default CourseTable;
