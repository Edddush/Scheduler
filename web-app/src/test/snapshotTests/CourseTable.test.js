import React from 'react';
import CourseTable from '../../components/CourseTable';
import renderer from 'react-test-renderer';


const columns = [
    { header: "name", accessProperty: "user_name" },
    { header: "DOB", accessProperty: "user_DOB" }
];

const callbackCol = {
    header: "Send birthday email",
    text: "Send",
    callback: () => { }
}

const data = [
    {
        user_name: "John",
        user_DOB: "xx-yy-2022"
    },
    {
        user_name: "Maya",
        user_DOB: "xx-yy-2021"
    },
    {
        user_name: "Sam",
        user_DOB: "xx-yy-2020"
    },
    {
        user_name: "Fred",
        user_DOB: "xx-yy-2027"
    },
    {
        user_name: "Jacob",
        user_DOB: "xx-yy-2029"
    }
]

it('renders with array of data', () => {
    const tree = renderer
        .create(
                <CourseTable columns={columns}
                    callbackCol={callbackCol}
                    data={data} />
        )
        .toJSON();
    expect(tree).toMatchSnapshot();
});

it('renders with array of no data', () => {
    const tree = renderer
        .create(<CourseTable columns={columns}
            callbackCol={callbackCol}
            data={[]} />)
        .toJSON();
    expect(tree).toMatchSnapshot();
});