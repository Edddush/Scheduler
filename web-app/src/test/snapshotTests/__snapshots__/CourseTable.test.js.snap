// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`renders with array of data 1`] = `
<div
  className="MuiPaper-root MuiPaper-elevation MuiPaper-rounded MuiPaper-elevation1 MuiTableContainer-root css-11xur9t-MuiPaper-root-MuiTableContainer-root"
  style={
    {
      "height": "400px",
      "overflowY": "scroll",
    }
  }
>
  <table
    aria-label="simple table"
    className="MuiTable-root MuiTable-stickyHeader css-1nfbgt6-MuiTable-root"
    role={null}
    style={
      {
        "overflow": "scroll",
      }
    }
  >
    <thead
      className="MuiTableHead-root css-15wwp11-MuiTableHead-root"
      role={null}
    >
      <tr
        className="MuiTableRow-root MuiTableRow-head css-1q1u3t4-MuiTableRow-root"
        role={null}
      >
        <th
          aria-sort={null}
          className="MuiTableCell-root MuiTableCell-head MuiTableCell-stickyHeader MuiTableCell-alignCenter MuiTableCell-sizeMedium css-dwuj3p-MuiTableCell-root"
          scope="col"
        >
          name
        </th>
        <th
          aria-sort={null}
          className="MuiTableCell-root MuiTableCell-head MuiTableCell-stickyHeader MuiTableCell-alignCenter MuiTableCell-sizeMedium css-dwuj3p-MuiTableCell-root"
          scope="col"
        >
          DOB
        </th>
        <th
          aria-sort={null}
          className="MuiTableCell-root MuiTableCell-head MuiTableCell-stickyHeader MuiTableCell-alignCenter MuiTableCell-sizeMedium css-dwuj3p-MuiTableCell-root"
          scope="col"
        >
          Send birthday email
        </th>
      </tr>
    </thead>
    <tbody
      className="MuiTableBody-root css-apqrd9-MuiTableBody-root"
      role={null}
    >
      <tr
        className="MuiTableRow-root css-34nofg-MuiTableRow-root"
        role={null}
      >
        <td
          aria-sort={null}
          className="MuiTableCell-root MuiTableCell-body MuiTableCell-alignCenter MuiTableCell-sizeMedium css-1yhpg23-MuiTableCell-root"
        >
          John
        </td>
        <td
          aria-sort={null}
          className="MuiTableCell-root MuiTableCell-body MuiTableCell-alignCenter MuiTableCell-sizeMedium css-1yhpg23-MuiTableCell-root"
        >
          xx-yy-2022
        </td>
        <td
          aria-sort={null}
          className="MuiTableCell-root MuiTableCell-body MuiTableCell-alignCenter MuiTableCell-sizeMedium css-1yhpg23-MuiTableCell-root"
        >
          <button
            className="btn btn-dark btn-lg"
            disabled={false}
            id="temp-button"
            onClick={[Function]}
            type="button"
          >
            Send
          </button>
        </td>
      </tr>
      <tr
        className="MuiTableRow-root css-34nofg-MuiTableRow-root"
        role={null}
      >
        <td
          aria-sort={null}
          className="MuiTableCell-root MuiTableCell-body MuiTableCell-alignCenter MuiTableCell-sizeMedium css-1yhpg23-MuiTableCell-root"
        >
          Maya
        </td>
        <td
          aria-sort={null}
          className="MuiTableCell-root MuiTableCell-body MuiTableCell-alignCenter MuiTableCell-sizeMedium css-1yhpg23-MuiTableCell-root"
        >
          xx-yy-2021
        </td>
        <td
          aria-sort={null}
          className="MuiTableCell-root MuiTableCell-body MuiTableCell-alignCenter MuiTableCell-sizeMedium css-1yhpg23-MuiTableCell-root"
        >
          <button
            className="btn btn-dark btn-lg"
            disabled={false}
            id="temp-button"
            onClick={[Function]}
            type="button"
          >
            Send
          </button>
        </td>
      </tr>
      <tr
        className="MuiTableRow-root css-34nofg-MuiTableRow-root"
        role={null}
      >
        <td
          aria-sort={null}
          className="MuiTableCell-root MuiTableCell-body MuiTableCell-alignCenter MuiTableCell-sizeMedium css-1yhpg23-MuiTableCell-root"
        >
          Sam
        </td>
        <td
          aria-sort={null}
          className="MuiTableCell-root MuiTableCell-body MuiTableCell-alignCenter MuiTableCell-sizeMedium css-1yhpg23-MuiTableCell-root"
        >
          xx-yy-2020
        </td>
        <td
          aria-sort={null}
          className="MuiTableCell-root MuiTableCell-body MuiTableCell-alignCenter MuiTableCell-sizeMedium css-1yhpg23-MuiTableCell-root"
        >
          <button
            className="btn btn-dark btn-lg"
            disabled={false}
            id="temp-button"
            onClick={[Function]}
            type="button"
          >
            Send
          </button>
        </td>
      </tr>
      <tr
        className="MuiTableRow-root css-34nofg-MuiTableRow-root"
        role={null}
      >
        <td
          aria-sort={null}
          className="MuiTableCell-root MuiTableCell-body MuiTableCell-alignCenter MuiTableCell-sizeMedium css-1yhpg23-MuiTableCell-root"
        >
          Fred
        </td>
        <td
          aria-sort={null}
          className="MuiTableCell-root MuiTableCell-body MuiTableCell-alignCenter MuiTableCell-sizeMedium css-1yhpg23-MuiTableCell-root"
        >
          xx-yy-2027
        </td>
        <td
          aria-sort={null}
          className="MuiTableCell-root MuiTableCell-body MuiTableCell-alignCenter MuiTableCell-sizeMedium css-1yhpg23-MuiTableCell-root"
        >
          <button
            className="btn btn-dark btn-lg"
            disabled={false}
            id="temp-button"
            onClick={[Function]}
            type="button"
          >
            Send
          </button>
        </td>
      </tr>
      <tr
        className="MuiTableRow-root css-34nofg-MuiTableRow-root"
        role={null}
      >
        <td
          aria-sort={null}
          className="MuiTableCell-root MuiTableCell-body MuiTableCell-alignCenter MuiTableCell-sizeMedium css-1yhpg23-MuiTableCell-root"
        >
          Jacob
        </td>
        <td
          aria-sort={null}
          className="MuiTableCell-root MuiTableCell-body MuiTableCell-alignCenter MuiTableCell-sizeMedium css-1yhpg23-MuiTableCell-root"
        >
          xx-yy-2029
        </td>
        <td
          aria-sort={null}
          className="MuiTableCell-root MuiTableCell-body MuiTableCell-alignCenter MuiTableCell-sizeMedium css-1yhpg23-MuiTableCell-root"
        >
          <button
            className="btn btn-dark btn-lg"
            disabled={false}
            id="temp-button"
            onClick={[Function]}
            type="button"
          >
            Send
          </button>
        </td>
      </tr>
    </tbody>
  </table>
</div>
`;

exports[`renders with array of no data 1`] = `
<div
  className="MuiPaper-root MuiPaper-elevation MuiPaper-rounded MuiPaper-elevation1 MuiTableContainer-root css-11xur9t-MuiPaper-root-MuiTableContainer-root"
  style={
    {
      "height": "400px",
      "overflowY": "scroll",
    }
  }
>
  <table
    aria-label="simple table"
    className="MuiTable-root MuiTable-stickyHeader css-1nfbgt6-MuiTable-root"
    role={null}
    style={
      {
        "overflow": "scroll",
      }
    }
  >
    <thead
      className="MuiTableHead-root css-15wwp11-MuiTableHead-root"
      role={null}
    >
      <tr
        className="MuiTableRow-root MuiTableRow-head css-1q1u3t4-MuiTableRow-root"
        role={null}
      >
        <th
          aria-sort={null}
          className="MuiTableCell-root MuiTableCell-head MuiTableCell-stickyHeader MuiTableCell-alignCenter MuiTableCell-sizeMedium css-dwuj3p-MuiTableCell-root"
          scope="col"
        >
          name
        </th>
        <th
          aria-sort={null}
          className="MuiTableCell-root MuiTableCell-head MuiTableCell-stickyHeader MuiTableCell-alignCenter MuiTableCell-sizeMedium css-dwuj3p-MuiTableCell-root"
          scope="col"
        >
          DOB
        </th>
        <th
          aria-sort={null}
          className="MuiTableCell-root MuiTableCell-head MuiTableCell-stickyHeader MuiTableCell-alignCenter MuiTableCell-sizeMedium css-dwuj3p-MuiTableCell-root"
          scope="col"
        >
          Send birthday email
        </th>
      </tr>
    </thead>
    <tbody
      className="MuiTableBody-root css-apqrd9-MuiTableBody-root"
      role={null}
    />
  </table>
</div>
`;
